diff --git a/src/function.cpp b/src/function.cpp
index 92fde81..74ed9e0 100644
--- a/src/function.cpp
+++ b/src/function.cpp
@@ -145,7 +145,7 @@ static bool loadProduction(const char *pData)
 {
 	PRODUCTION_FUNCTION	*psFunction;
 	char					functionName[MAX_STR_LENGTH], bodySize[MAX_STR_LENGTH];
-	UDWORD					productionOutput;
+	uint32_t					productionOutput;
 
 	psFunction = (PRODUCTION_FUNCTION *)malloc(sizeof(PRODUCTION_FUNCTION));
 	memset(psFunction, 0, sizeof(PRODUCTION_FUNCTION));
@@ -162,7 +162,7 @@ static bool loadProduction(const char *pData)
 	//read the data in
 	functionName[0] = '\0';
 	bodySize[0] = '\0';
-	sscanf(pData, "%255[^,'\r\n],%255[^,'\r\n],%d", functionName, bodySize,
+	sscanf(pData, "%255[^,'\r\n],%255[^,'\r\n],%u", functionName, bodySize,
 	       &productionOutput);
 
 	//allocate storage for the name
@@ -174,15 +174,14 @@ static bool loadProduction(const char *pData)
 		return false;
 	}
 
-	//check prod output < UWORD_MAX
-	if (productionOutput < UWORD_MAX)
+	//check prod output <= UINT32_MAX
+	if (productionOutput <= UINT32_MAX)
 	{
-		psFunction->productionOutput = (UWORD)productionOutput;
+		psFunction->productionOutput = (uint32_t)productionOutput;
 	}
 	else
 	{
-		ASSERT(false, "loadProduction: production Output too big for %s", psFunction->pName);
-
+		ASSERT(false, "loadProduction: production Output too big for %s, production Output set to 0", psFunction->pName);
 		psFunction->productionOutput = 0;
 	}
 
@@ -960,7 +959,7 @@ void structureProductionUpgrade(STRUCTURE *psBuilding)
 {
 	FACTORY						*pFact;
 	PRODUCTION_FUNCTION			*pFactFunc;
-	UDWORD						type, baseOutput, i;
+	uint32_t						type, baseOutput, i;
 	STRUCTURE_STATS             *psStat;
 
 	switch (psBuilding->pStructureType->type)
@@ -999,7 +998,7 @@ void structureProductionUpgrade(STRUCTURE *psBuilding)
 		}
 	}
 
-	pFact->productionOutput = (UBYTE)(baseOutput + (pFactFunc->productionOutput *
+	pFact->productionOutput = (uint32_t)(baseOutput + (pFactFunc->productionOutput *
 	        asProductionUpgrade[psBuilding->player][type].modifier) / 100);
 }
 
diff --git a/src/functiondef.h b/src/functiondef.h
index eb87abb..244b661 100644
--- a/src/functiondef.h
+++ b/src/functiondef.h
@@ -112,7 +112,7 @@ struct PRODUCTION_UPGRADE_FUNCTION : public FUNCTION
 struct PRODUCTION_FUNCTION : public FUNCTION
 {
 	BODY_SIZE capacity;      // The max size of body the factory can produce
-	UWORD productionOutput;	 // Droid Build Points Produced Per Build Cycle
+	uint32_t productionOutput;	 // Droid Build Points Produced Per Build Cycle
 };
 
 /*To research topics available*/
diff --git a/src/intdisplay.cpp b/src/intdisplay.cpp
index 995978e..2cabd71 100644
--- a/src/intdisplay.cpp
+++ b/src/intdisplay.cpp
@@ -267,7 +267,7 @@ void intUpdateProgressBar(WIDGET *psWidget, W_CONTEXT *psContext)
 				{
 					// Started production. Set the colour of the bar to yellow.
 					int buildPointsTotal = FactoryGetTemplate(Manufacture)->buildPoints;
-					int buildRate = Manufacture->timeStartHold == 0? Manufacture->productionOutput : 0;
+					uint32_t buildRate = Manufacture->timeStartHold == 0? Manufacture->productionOutput : 0;
 					formatTime(BarGraph, buildPointsTotal - Manufacture->buildPointsRemaining, buildPointsTotal, buildRate, _("Construction Progress"));
 				}
 				else
diff --git a/src/structure.cpp b/src/structure.cpp
index 82ac814..d83b574 100644
--- a/src/structure.cpp
+++ b/src/structure.cpp
@@ -1708,12 +1708,11 @@ STRUCTURE* buildStructureDir(STRUCTURE_STATS *pStructureType, UDWORD x, UDWORD y
 
 				//quick check not trying to add too much
 				ASSERT_OR_RETURN(NULL, psBuilding->pFunctionality->factory.productionOutput +
-					((PRODUCTION_FUNCTION*)pStructureType->asFuncList[0])->productionOutput < UBYTE_MAX,
+					((PRODUCTION_FUNCTION*)pStructureType->asFuncList[0])->productionOutput <= UINT32_MAX,
 					"building factory module - production Output is too big");
 
 				psBuilding->pFunctionality->factory.productionOutput += ((
 					PRODUCTION_FUNCTION*)pStructureType->asFuncList[0])->productionOutput;
-
 				capacity = psBuilding->pFunctionality->factory.capacity;
 			}
 		}
@@ -1936,7 +1935,7 @@ static bool setFunctionality(STRUCTURE	*psBuilding, STRUCTURE_TYPE functionType)
 			unsigned int x, y;
 
 			psFactory->capacity = (UBYTE) ((PRODUCTION_FUNCTION*)psBuilding->pStructureType->asFuncList[0])->capacity;
-			psFactory->productionOutput = (UBYTE) ((PRODUCTION_FUNCTION*)psBuilding->pStructureType->asFuncList[0])->productionOutput;
+			psFactory->productionOutput = ((PRODUCTION_FUNCTION*)psBuilding->pStructureType->asFuncList[0])->productionOutput;
 			psFactory->psSubject = NULL;
 
 			// Default the secondary order - AB 22/04/99
diff --git a/src/structuredef.h b/src/structuredef.h
index eeee1da..6c971d9 100644
--- a/src/structuredef.h
+++ b/src/structuredef.h
@@ -179,7 +179,7 @@ struct FACTORY
 											   can produce*/
 	uint8_t                         productionLoops;        ///< Number of loops to perform. Not synchronised, and only meaningful for selectedPlayer.
 	UBYTE				loopsPerformed;		/* how many times the loop has been performed*/
-	UBYTE				productionOutput;	/* Droid Build Points Produced Per
+	uint32_t				productionOutput;	/* Droid Build Points Produced Per
 											   Build Cycle*/
 	DROID_TEMPLATE *                psSubject;              ///< The subject the structure is working on.
 	DROID_TEMPLATE *                psSubjectPending;       ///< The subject the structure is going to working on. (Pending = not yet synchronised.)
