diff --git a/src/display.cpp b/src/display.cpp
index 4d57132..cd9b486 100644
--- a/src/display.cpp
+++ b/src/display.cpp
@@ -1776,10 +1776,11 @@ static void dealWithLMBDroid(DROID* psDroid, SELECTION_TYPE selection)
 				if (!psDroid->selected)
 				{
 					CONPRINTF(ConsoleString, (ConsoleString,
-					                          _("%s - Damage %d%% - Experience %.1f, %s"),
+								_("%s - Body Points: %d of %d - Experience %.1f, %s"),
-					                          droidGetName(psDroid),
+								droidGetName(psDroid),
-					                          100 - clip(PERCENT(psDroid->body, psDroid->originalBody), 0, 100),
+								psDroid->body,
+								psDroid->originalBody,
-					                          psDroid->experience / 65536.f, _(getDroidLevelName(psDroid))));
+								psDroid->experience / 65536.f, _(getDroidLevelName(psDroid))));
 
 					FeedbackOrderGiven();
 				}
@@ -1792,10 +1793,11 @@ static void dealWithLMBDroid(DROID* psDroid, SELECTION_TYPE selection)
 	else // Clicked on allied unit with no other possible actions
 	{
 		CONPRINTF(ConsoleString, (ConsoleString,
-		                          _("%s - Allied - Damage %d%% - Experience %d, %s"),
+					_("%s - Allied - Body Points: %d of %d - Experience %d, %s"),
-		                          droidGetName(psDroid),
+					droidGetName(psDroid),
-		                          100 - clip(PERCENT(psDroid->body, psDroid->originalBody), 0, 100),
+					psDroid->body,
+					psDroid->originalBody,
-		                          psDroid->experience / 65536, getDroidLevelName(psDroid)));
+					psDroid->experience / 65536, getDroidLevelName(psDroid)));
 
 		FeedbackOrderGiven();
 	}
@@ -2276,10 +2278,11 @@ static void dealWithRMB( void )
								if (!psDroid->selected)
								{
									CONPRINTF(ConsoleString, (ConsoleString,
-									                          _("%s - Damage %d%% - Experience %.1f, %s"),
+												_("%s - Body Points: %d of %d - Experience %.1f, %s"),
-									                          droidGetName(psDroid),
+												droidGetName(psDroid),
-									                          100 - clip(PERCENT(psDroid->body, psDroid->originalBody), 0, 100),
+												psDroid->body,
+												psDroid->originalBody,
-									                          psDroid->experience / 65536.f, _(getDroidLevelName(psDroid))));
+												psDroid->experience/65536.f, _(getDroidLevelName(psDroid))));

									FeedbackOrderGiven();
								}
diff --git a/src/structure.cpp b/src/structure.cpp
index ed0159d..d79fef9 100644
--- a/src/structure.cpp
+++ b/src/structure.cpp
@@ -5638,8 +5638,8 @@ void printStructureInfo(STRUCTURE *psStructure)
 		{
 			unsigned int assigned_droids = countAssignedDroids(psStructure);
 
-			CONPRINTF(ConsoleString, (ConsoleString, ngettext("%s - %u Unit assigned - Damage %3.0f%%", "%s - %u Units assigned - Damage %3.0f%%", assigned_droids),
-			                          getStatName(psStructure->pStructureType), assigned_droids, getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, ngettext("%s - %u Unit assigned - Body Points: %d of %d", "%s - %u Units assigned - Body Points: %d of %d", assigned_droids),
+					  getStatName(psStructure->pStructureType), assigned_droids, psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	case REF_DEFENSE:
@@ -5664,13 +5664,13 @@ void printStructureInfo(STRUCTURE *psStructure)
 			{
 				unsigned int assigned_droids = countAssignedDroids(psStructure);
 
-				CONPRINTF(ConsoleString, (ConsoleString, ngettext("%s - %u Unit assigned - Damage %3.0f%%", "%s - %u Units assigned - Damage %3.0f%%", assigned_droids),
-				                          getStatName(psStructure->pStructureType), assigned_droids, getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, ngettext("%s - %u Unit assigned - Body Points: %d of %d", "%s - %u Units assigned - Body Points: %d of %d", assigned_droids),
+									getStatName(psStructure->pStructureType), assigned_droids, psStructure->body, structureBodyBuilt(psStructure)));
 			}
 			else
 			{
-				CONPRINTF(ConsoleString, (ConsoleString, _("%s - Damage %3.0f%%"),
-				                          getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Body Points: %d of %d"),
+									getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 			}
 		break;
 	case REF_REPAIR_FACILITY:
@@ -5683,7 +5683,7 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, "%s - Damage %3.0f%%", getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, "%s - Body Points: %d of %d", getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	case REF_RESOURCE_EXTRACTOR:
@@ -5696,7 +5696,7 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, "%s - Damage %3.0f%%", getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, "%s - Body Points: %d of %d", getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	case REF_POWER_GEN:
@@ -5719,8 +5719,8 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Connected %u of %u - Damage %3.0f%%"),
-			                          getStatName(psStructure->pStructureType), numConnected, NUM_POWER_MODULES, getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Connected %u of %u - Body Points: %d of %d"),
+									getStatName(psStructure->pStructureType), numConnected, NUM_POWER_MODULES, psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	case REF_CYBORG_FACTORY:
@@ -5737,8 +5737,8 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Damage %3.0f%%"),
-			                          getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Body Points: %d of %d"),
+									getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	case REF_RESEARCH:
@@ -5752,8 +5752,8 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Damage %3.0f%%"),
-			                          getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Body Points: %d of %d"),
+									getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	default:
@@ -5766,8 +5766,8 @@ void printStructureInfo(STRUCTURE *psStructure)
 		else
 #endif
 		{
-			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Damage %3.0f%%"),
-			                          getStatName(psStructure->pStructureType), getStructureDamage(psStructure) * (100.f / 65536.f)));
+			CONPRINTF(ConsoleString, (ConsoleString, _("%s - Body Points: %d of %d"),
+									getStatName(psStructure->pStructureType), psStructure->body, structureBodyBuilt(psStructure)));
 		}
 		break;
 	}
