diff --git a/src/projectile.cpp b/src/projectile.cpp
index 972681a..52677f7 100644
--- a/src/projectile.cpp
+++ b/src/projectile.cpp
@@ -1240,7 +1240,7 @@ static void proj_PostImpactFunc(PROJECTILE *psObj)
 	int age = gameTime - psObj->born;
 
 	/* Time to finish postimpact effect? */
-	if (age > psStats->radiusLife && age > psStats->upgrade[psObj->player].periodicalDamageTime)
+	if (age >= psStats->radiusLife && age >= psStats->upgrade[psObj->player].periodicalDamageTime)
 	{
 		psObj->state = PROJ_INACTIVE;
 		return;
