diff --git a/src/move.cpp b/src/move.cpp
index 5f5c73d..a8fd727 100644
--- a/src/move.cpp
+++ b/src/move.cpp
@@ -116,7 +116,7 @@
 // How fast a wheeled droid decelerates
 #define WHEELED_SKID_DECEL		350
 // How fast a hover droid decelerates
-#define HOVER_SKID_DECEL		120
+#define HOVER_SKID_DECEL		200
 
 /************************************************************************************/
 /*             Person model defines                                                 */
