diff --git a/src/combat.cpp b/src/combat.cpp
index b227d1b..40c8649 100644
--- a/src/combat.cpp
+++ b/src/combat.cpp
@@ -371,7 +371,7 @@ void counterBatteryFire(BASE_OBJECT *psAttacker, BASE_OBJECT *psTarget)
  */
 int32_t objDamage(BASE_OBJECT *psObj, unsigned damage, unsigned originalhp, WEAPON_CLASS weaponClass, WEAPON_SUBCLASS weaponSubClass, bool isDamagePerSecond, int minDamage)
 {
-	int	actualDamage, level = 1, lastHit = psObj->timeLastHit;
+	int	actualDamage, level = 0, lastHit = psObj->timeLastHit;
 	int armour = objArmour(psObj, weaponClass);
 
 	// If the previous hit was by an EMP cannon and this one is not:
@@ -461,7 +461,7 @@ int32_t objDamage(BASE_OBJECT *psObj, unsigned damage, unsigned originalhp, WEAP
 unsigned int objGuessFutureDamage(WEAPON_STATS *psStats, unsigned int player, BASE_OBJECT *psTarget)
 {
 	unsigned int damage;
-	int actualDamage, armour, level = 1;
+	int	actualDamage, armour = 0, level = 0;
 
 	if (psTarget == NULL)
 	{
