diff --git a/src/move.cpp b/src/move.cpp
index b310ee5..d4793a9 100644
--- a/src/move.cpp
+++ b/src/move.cpp
@@ -57,9 +57,9 @@
 #include "qtscript.h"
 
 /* max and min vtol heights above terrain */
-#define	VTOL_HEIGHT_MIN				250
-#define	VTOL_HEIGHT_LEVEL			300
-#define	VTOL_HEIGHT_MAX				350
+#define	VTOL_HEIGHT_MIN				400
+#define	VTOL_HEIGHT_LEVEL			450
+#define	VTOL_HEIGHT_MAX				500
 
 // Maximum size of an object for collision
 #define OBJ_MAXRADIUS	(TILE_UNITS * 4)
