diff --git a/src/projectile.cpp b/src/projectile.cpp
index b2d8b4d..11f6d3d 100644
--- a/src/projectile.cpp
+++ b/src/projectile.cpp
@@ -1421,8 +1421,7 @@ static void proj_checkBurnDamage(PROJECTILE *psProj)
 			psCurr = *it;
 			if(isDead(psCurr))
 			{
-				psProj->burnVictims.erase(it);
-				break;	// needed, because we are not resetting the iterator
+				it = psProj->burnVictims.erase(it);
 			}
 		}
 	}
