diff --git a/src/combat.cpp b/src/combat.cpp
index 7b59e28..33e0bcd 100644
--- a/src/combat.cpp
+++ b/src/combat.cpp
@@ -282,10 +282,10 @@ bool combFire(WEAPON *psWeap, BASE_OBJECT *psAttacker, BASE_OBJECT *psTarget, in
 	}
 	else /* Deal with a missed shot */
 	{
-		const int minOffset = 5;
+		const int minOffset = 50;
 
 		int missDist = 2 * (100 - resultHitChance) + minOffset;
-		Vector3i miss = Vector3i(iSinCosR(gameRand(DEG(360)), missDist), 0);
+		Vector3i miss = Vector3i(iSinCosR(gameRand(DEG(360)), missDist), gameRand(128) - 64);
 		predict += miss;
 
 		psTarget = NULL;  // Missed the target, so don't expect to hit it.
