diff --git a/configure.ac b/configure.ac
index 226b4bf..2d0dd8e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -123,7 +123,7 @@ case ${host_os} in
 	*mingw32*)
 		host_os_mingw32=yes
 		AC_CHECK_TOOL([WINDRES], [windres], AC_MSG_ERROR([windres not found]))
-		WIN32_LIBS='-ldbghelp -lshfolder -lshlwapi -lpsapi -lshell32 -lwinmm -lws2_32 -liphlpapi'
+		WIN32_LIBS='-ldbghelp -lshfolder -lshlwapi -lpsapi -lshell32 -lwinmm -lws2_32 -liphlpapi -lwtsapi32'
 		AC_SUBST([WIN32_LIBS], [${WIN32_LIBS}])
 		LDFLAGS="-Wl,-subsystem,windows ${LDFLAGS}"
 		;;
diff --git a/pkg/nsis/warzone2100.nsi b/pkg/nsis/warzone2100.nsi
index 12dbdac..8c278b7 100644
--- a/pkg/nsis/warzone2100.nsi
+++ b/pkg/nsis/warzone2100.nsi
@@ -621,15 +621,15 @@ FunctionEnd
 ;  LangString DESC_SecOriginalMod ${LANG_DUTCH} "Speel het spel met de originele 1.10 versie balans stats."
 
   ;German
-  LangString WZWelcomeText ${LANG_GERMAN} "Dieser Wizard wird Sie durch die Warzone-2100-Installation fьhren.\r\n\r\nEs wird empfohlen sдmtliche anderen Anwendungen zu schlieЯen, bevor Sie das Setup starten. Dies ermцglicht es relevante Systemdateien zu aktualisieren, ohne neustarten zu mьssen.\r\n\r\nWarzone 2100 ist zu 100% kostenlos, falls Sie dafьr gezahlt haben, lassen Sie es uns wissen!\r\n\r\nKlicken Sie auf Weiter, um fortzufahren."
-  LangString WZ_GPL_NEXT ${LANG_GERMAN} "nдchste"
+  LangString WZWelcomeText ${LANG_GERMAN} "Dieser Wizard wird Sie durch die Warzone-2100-Installation fГјhren.\r\n\r\nEs wird empfohlen sГ¤mtliche anderen Anwendungen zu schlieГџen, bevor Sie das Setup starten. Dies ermГ¶glicht es relevante Systemdateien zu aktualisieren, ohne neustarten zu mГјssen.\r\n\r\nWarzone 2100 ist zu 100% kostenlos, falls Sie dafГјr gezahlt haben, lassen Sie es uns wissen!\r\n\r\nKlicken Sie auf Weiter, um fortzufahren."
+  LangString WZ_GPL_NEXT ${LANG_GERMAN} "nГ¤chste"
 
 
   LangString TEXT_SecBase ${LANG_GERMAN} "Core files"
-  LangString DESC_SecBase ${LANG_GERMAN} "Die Kerndateien, die fьr Warzone 2100 benцtigt werden."
+  LangString DESC_SecBase ${LANG_GERMAN} "Die Kerndateien, die fГјr Warzone 2100 benГ¶tigt werden."
 
   LangString TEXT_SecOpenAL ${LANG_GERMAN} "OpenAL Bibliotheken"
-  LangString DESC_SecOpenAL ${LANG_GERMAN} "Bibliotheken fьr OpenAL, ein freies Audio Interface. Implementation von Creative Labs."
+  LangString DESC_SecOpenAL ${LANG_GERMAN} "Bibliotheken fГјr OpenAL, ein freies Audio Interface. Implementation von Creative Labs."
 
 ;  LangString TEXT_SecMods ${LANG_GERMAN} "Mods"
 ;  LangString DESC_SecMods ${LANG_GERMAN} "Verschiedene Mods."
@@ -638,63 +638,63 @@ FunctionEnd
   LangString DESC_SecFMVs ${LANG_GERMAN} "Videos herunterladen und installieren."
 
   LangString TEXT_SecFMVs_EngHi ${LANG_GERMAN} "English (HQ)"
-  LangString DESC_SecFMVs_EngHi ${LANG_GERMAN} "Videos in besserer Qualitдt und auf Englisch herunterladen und installieren (920 MiB)."
+  LangString DESC_SecFMVs_EngHi ${LANG_GERMAN} "Videos in besserer QualitГ¤t und auf Englisch herunterladen und installieren (920 MiB)."
 
   LangString TEXT_SecFMVs_Eng ${LANG_GERMAN} "English"
   LangString DESC_SecFMVs_Eng ${LANG_GERMAN} "Die englischen Videos herunterladen und installieren (545 MiB)."
 
   LangString TEXT_SecFMVs_EngLo ${LANG_GERMAN} "English (LQ)"
-  LangString DESC_SecFMVs_EngLo ${LANG_GERMAN} "Die englischen Videos in geringer Qualitдt herunterladen und installieren (162 MiB)."
+  LangString DESC_SecFMVs_EngLo ${LANG_GERMAN} "Die englischen Videos in geringer QualitГ¤t herunterladen und installieren (162 MiB)."
 
   LangString TEXT_SecFMVs_Ger ${LANG_GERMAN} "German"
   LangString DESC_SecFMVs_Ger ${LANG_GERMAN} "Die deutschen Videos herunterladen und installieren (460 MiB)."
 
   LangString TEXT_SecNLS ${LANG_GERMAN} "Language files"
-  LangString DESC_SecNLS ${LANG_GERMAN} "Unterstьtzung fьr Sprachen auЯer Englisch (Deutsch inbegriffen)."
+  LangString DESC_SecNLS ${LANG_GERMAN} "UnterstГјtzung fГјr Sprachen auГџer Englisch (Deutsch inbegriffen)."
 
   LangString TEXT_SecNLS_WinFonts ${LANG_GERMAN} "WinFonts"
-  LangString DESC_SecNLS_WinFonts ${LANG_GERMAN} "Den Windows-Schriftarten-Ordner in den Suchpfad aufnehmen. Nutzen Sie dies, falls Sie spдter eigene Schriftarten in der Konfigurationsdatei eingeben wollen oder es zu Problemen mit der Standardschriftart kommt. Kann unter Vista und spдter langsam sein!"
+  LangString DESC_SecNLS_WinFonts ${LANG_GERMAN} "Den Windows-Schriftarten-Ordner in den Suchpfad aufnehmen. Nutzen Sie dies, falls Sie spГ¤ter eigene Schriftarten in der Konfigurationsdatei eingeben wollen oder es zu Problemen mit der Standardschriftart kommt. Kann unter Vista und spГ¤ter langsam sein!"
 
   LangString TEXT_SecOriginalMod ${LANG_GERMAN} "1.10 balance"
   LangString DESC_SecOriginalMod ${LANG_GERMAN} "Spielen Sie das Spiel mit dem Balancing aus der Originalversion 1.10."
 
   ;Russian
-  LangString WZWelcomeText ${LANG_RUSSIAN} "Этот помощник установки поможет вам установить Warzone2100.\r\n\r\nПеред началом рекомендуем закрыть все другие приложения. Это позволит обновить соответствующие системные файлы без перезагрузки системы.\r\n\r\nWarzone2100 100% бесплатный, если вы за него заплатили сообщите нам!\r\n\r\nНажмите Далее для продолжения."
-  LangString WZ_GPL_NEXT ${LANG_RUSSIAN} "Согласен"
+  LangString WZWelcomeText ${LANG_RUSSIAN} "Р­С‚РѕС‚ РїРѕРјРѕС‰РЅРёРє СѓСЃС‚Р°РЅРѕРІРєРё РїРѕРјРѕР¶РµС‚ РІР°Рј СѓСЃС‚Р°РЅРѕРІРёС‚СЊ Warzone2100.\r\n\r\nРџРµСЂРµРґ РЅР°С‡Р°Р»РѕРј СЂРµРєРѕРјРµРЅРґСѓРµРј Р·Р°РєСЂС‹С‚СЊ РІСЃРµ РґСЂСѓРіРёРµ РїСЂРёР»РѕР¶РµРЅРёСЏ. Р­С‚Рѕ РїРѕР·РІРѕР»РёС‚ РѕР±РЅРѕРІРёС‚СЊ СЃРѕРѕС‚РІРµС‚СЃС‚РІСѓСЋС‰РёРµ СЃРёСЃС‚РµРјРЅС‹Рµ С„Р°Р№Р»С‹ Р±РµР· РїРµСЂРµР·Р°РіСЂСѓР·РєРё СЃРёСЃС‚РµРјС‹.\r\n\r\nWarzone2100 100% Р±РµСЃРїР»Р°С‚РЅС‹Р№, РµСЃР»Рё РІС‹ Р·Р° РЅРµРіРѕ Р·Р°РїР»Р°С‚РёР»Рё СЃРѕРѕР±С‰РёС‚Рµ РЅР°Рј!\r\n\r\nРќР°Р¶РјРёС‚Рµ Р”Р°Р»РµРµ РґР»СЏ РїСЂРѕРґРѕР»Р¶РµРЅРёСЏ."
+  LangString WZ_GPL_NEXT ${LANG_RUSSIAN} "РЎРѕРіР»Р°СЃРµРЅ"
 
 
-  LangString TEXT_SecBase ${LANG_RUSSIAN} "Базовые файлы"
-  LangString DESC_SecBase ${LANG_RUSSIAN} "Файлы требуемые для запуска Warzone 2100."
+  LangString TEXT_SecBase ${LANG_RUSSIAN} "Р‘Р°Р·РѕРІС‹Рµ С„Р°Р№Р»С‹"
+  LangString DESC_SecBase ${LANG_RUSSIAN} "Р¤Р°Р№Р»С‹ С‚СЂРµР±СѓРµРјС‹Рµ РґР»СЏ Р·Р°РїСѓСЃРєР° Warzone 2100."
 
-  LangString TEXT_SecOpenAL ${LANG_RUSSIAN} "Библиотека OpenAL"
-  LangString DESC_SecOpenAL ${LANG_RUSSIAN} "Свободно распространяемый аппаратно- программный интерфейс (API) для работы с аудиоданными. Версия от Creative Labs."
+  LangString TEXT_SecOpenAL ${LANG_RUSSIAN} "Р‘РёР±Р»РёРѕС‚РµРєР° OpenAL"
+  LangString DESC_SecOpenAL ${LANG_RUSSIAN} "РЎРІРѕР±РѕРґРЅРѕ СЂР°СЃРїСЂРѕСЃС‚СЂР°РЅСЏРµРјС‹Р№ Р°РїРїР°СЂР°С‚РЅРѕ- РїСЂРѕРіСЂР°РјРјРЅС‹Р№ РёРЅС‚РµСЂС„РµР№СЃ (API) РґР»СЏ СЂР°Р±РѕС‚С‹ СЃ Р°СѓРґРёРѕРґР°РЅРЅС‹РјРё. Р’РµСЂСЃРёСЏ РѕС‚ Creative Labs."
 
-;  LangString TEXT_SecMods ${LANG_RUSSIAN} "Модификации"
-;  LangString DESC_SecMods ${LANG_RUSSIAN} "Различные модификации для Warzone 2100."
+;  LangString TEXT_SecMods ${LANG_RUSSIAN} "РњРѕРґРёС„РёРєР°С†РёРё"
+;  LangString DESC_SecMods ${LANG_RUSSIAN} "Р Р°Р·Р»РёС‡РЅС‹Рµ РјРѕРґРёС„РёРєР°С†РёРё РґР»СЏ Warzone 2100."
 
-  LangString TEXT_SecFMVs ${LANG_RUSSIAN} "Видео"
-  LangString DESC_SecFMVs ${LANG_RUSSIAN} "Скачать и установить внутриигровые ролики."
+  LangString TEXT_SecFMVs ${LANG_RUSSIAN} "Р’РёРґРµРѕ"
+  LangString DESC_SecFMVs ${LANG_RUSSIAN} "РЎРєР°С‡Р°С‚СЊ Рё СѓСЃС‚Р°РЅРѕРІРёС‚СЊ РІРЅСѓС‚СЂРёРёРіСЂРѕРІС‹Рµ СЂРѕР»РёРєРё."
 
-  LangString TEXT_SecFMVs_EngHi ${LANG_RUSSIAN} "Английские (HQ)"
+  LangString TEXT_SecFMVs_EngHi ${LANG_RUSSIAN} "РђРЅРіР»РёР№СЃРєРёРµ (HQ)"
   LangString DESC_SecFMVs_EngHi ${LANG_RUSSIAN} "Download and install higher-quality English in-game cutscenes (920 MB)."
 
-  LangString TEXT_SecFMVs_Eng ${LANG_RUSSIAN} "Английские"
-  LangString DESC_SecFMVs_Eng ${LANG_RUSSIAN} "Скачать и установить внутриигровые ролики на английском языке (545 MB)."
+  LangString TEXT_SecFMVs_Eng ${LANG_RUSSIAN} "РђРЅРіР»РёР№СЃРєРёРµ"
+  LangString DESC_SecFMVs_Eng ${LANG_RUSSIAN} "РЎРєР°С‡Р°С‚СЊ Рё СѓСЃС‚Р°РЅРѕРІРёС‚СЊ РІРЅСѓС‚СЂРёРёРіСЂРѕРІС‹Рµ СЂРѕР»РёРєРё РЅР° Р°РЅРіР»РёР№СЃРєРѕРј СЏР·С‹РєРµ (545 MB)."
 
-  LangString TEXT_SecFMVs_EngLo ${LANG_RUSSIAN} "Английские (LQ)"
-  LangString DESC_SecFMVs_EngLo ${LANG_RUSSIAN} "Скачать и установить внутриигровые ролики (низкого качества) на английском языке (162 MB)."
+  LangString TEXT_SecFMVs_EngLo ${LANG_RUSSIAN} "РђРЅРіР»РёР№СЃРєРёРµ (LQ)"
+  LangString DESC_SecFMVs_EngLo ${LANG_RUSSIAN} "РЎРєР°С‡Р°С‚СЊ Рё СѓСЃС‚Р°РЅРѕРІРёС‚СЊ РІРЅСѓС‚СЂРёРёРіСЂРѕРІС‹Рµ СЂРѕР»РёРєРё (РЅРёР·РєРѕРіРѕ РєР°С‡РµСЃС‚РІР°) РЅР° Р°РЅРіР»РёР№СЃРєРѕРј СЏР·С‹РєРµ (162 MB)."
 
-  LangString TEXT_SecFMVs_Ger ${LANG_RUSSIAN} "Немецкие"
-  LangString DESC_SecFMVs_Ger ${LANG_RUSSIAN} "Скачать и установить внутриигровые ролики на немецком языке (460 MB)."
+  LangString TEXT_SecFMVs_Ger ${LANG_RUSSIAN} "РќРµРјРµС†РєРёРµ"
+  LangString DESC_SecFMVs_Ger ${LANG_RUSSIAN} "РЎРєР°С‡Р°С‚СЊ Рё СѓСЃС‚Р°РЅРѕРІРёС‚СЊ РІРЅСѓС‚СЂРёРёРіСЂРѕРІС‹Рµ СЂРѕР»РёРєРё РЅР° РЅРµРјРµС†РєРѕРј СЏР·С‹РєРµ (460 MB)."
 
-  LangString TEXT_SecNLS ${LANG_RUSSIAN} "Языковые файлы"
-  LangString DESC_SecNLS ${LANG_RUSSIAN} "Поддержка Русского и других языков."
+  LangString TEXT_SecNLS ${LANG_RUSSIAN} "РЇР·С‹РєРѕРІС‹Рµ С„Р°Р№Р»С‹"
+  LangString DESC_SecNLS ${LANG_RUSSIAN} "РџРѕРґРґРµСЂР¶РєР° Р СѓСЃСЃРєРѕРіРѕ Рё РґСЂСѓРіРёС… СЏР·С‹РєРѕРІ."
 
-  LangString TEXT_SecNLS_WinFonts ${LANG_RUSSIAN} "WinШрифты"
-  LangString DESC_SecNLS_WinFonts ${LANG_RUSSIAN} "Задействовать папку шрифтов Windows при поиске. Помогает если есть проблемы с поставляемыми шрифтами. На Висте возможно замедление при загрузке!"
+  LangString TEXT_SecNLS_WinFonts ${LANG_RUSSIAN} "WinРЁСЂРёС„С‚С‹"
+  LangString DESC_SecNLS_WinFonts ${LANG_RUSSIAN} "Р—Р°РґРµР№СЃС‚РІРѕРІР°С‚СЊ РїР°РїРєСѓ С€СЂРёС„С‚РѕРІ Windows РїСЂРё РїРѕРёСЃРєРµ. РџРѕРјРѕРіР°РµС‚ РµСЃР»Рё РµСЃС‚СЊ РїСЂРѕР±Р»РµРјС‹ СЃ РїРѕСЃС‚Р°РІР»СЏРµРјС‹РјРё С€СЂРёС„С‚Р°РјРё. РќР° Р’РёСЃС‚Рµ РІРѕР·РјРѕР¶РЅРѕ Р·Р°РјРµРґР»РµРЅРёРµ РїСЂРё Р·Р°РіСЂСѓР·РєРµ!"
 
-  LangString TEXT_SecOriginalMod ${LANG_RUSSIAN} "Баланс 1.10"
-  LangString DESC_SecOriginalMod ${LANG_RUSSIAN} "Играть в игру с балансом от оригинальной версии 1.10."
+  LangString TEXT_SecOriginalMod ${LANG_RUSSIAN} "Р‘Р°Р»Р°РЅСЃ 1.10"
+  LangString DESC_SecOriginalMod ${LANG_RUSSIAN} "РРіСЂР°С‚СЊ РІ РёРіСЂСѓ СЃ Р±Р°Р»Р°РЅСЃРѕРј РѕС‚ РѕСЂРёРіРёРЅР°Р»СЊРЅРѕР№ РІРµСЂСЃРёРё 1.10."
 
   ;Assign language strings to sections
   !insertmacro MUI_FUNCTION_DESCRIPTION_BEGIN
diff --git a/src/Makefile.am b/src/Makefile.am
index ee4d490..4cbd87a 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -324,7 +324,7 @@ endif
 # Here, we need to bring in the "plugins" for Qt5 and they MUST come before the other QT libs.
 # FIXME: At a future date, we should check if they want static builds, but, since we only support those for now...
 if MINGW32
-SQT5_LIBS =  -L$(QT5_PLUGIN)/plugins/platforms -lQt5PlatformSupport -lqwindows
+SQT5_LIBS =  -L$(QT5_PLUGIN)/plugins/platforms -lqwindows -lQt5ThemeSupport -lQt5EventDispatcherSupport -lQt5WindowsUIAutomationSupport -lQt5FontDatabaseSupport
 else
 SQT5_LIBS =
 endif
