diff -ur src_3.1rc1/structure.cpp src/structure.cpp
--- src_3.1rc1/structure.cpp	2012-08-08 00:31:46.000000000 +0200
+++ src/structure.cpp	2012-08-25 13:27:41.000000000 +0200
@@ -5457,7 +5457,7 @@
 		prevState = intGetResearchState();
 	}
 
-	psBuilding->currentBuildPts = (SWORD)psBuilding->pStructureType->buildPoints;
+	psBuilding->currentBuildPts = psBuilding->pStructureType->buildPoints;
 	psBuilding->status = SS_BUILT;
 
 	visTilesUpdate(psBuilding);
diff -ur src_3.1rc1/structuredef.h src/structuredef.h
--- src_3.1rc1/structuredef.h	2012-08-08 00:31:46.000000000 +0200
+++ src/structuredef.h	2012-08-25 13:05:41.000000000 +0200
@@ -252,7 +252,7 @@
 
 	STRUCTURE_STATS     *pStructureType;            /* pointer to the structure stats for this type of building */
 	STRUCT_STATES       status;                     /* defines whether the structure is being built, doing nothing or performing a function */
-	SWORD               currentBuildPts;            /* the build points currently assigned to this structure */
+	int32_t             currentBuildPts;            /* the build points currently assigned to this structure */
 	SWORD               resistance;                 /* current resistance points, 0 = cannot be attacked electrically */
 	UDWORD              lastResistance;             /* time the resistance was last increased*/
 	FUNCTIONALITY       *pFunctionality;            /* pointer to structure that contains fields necessary for functionality */
